### PR TITLE
Add TransR Implementation

### PR DESCRIPTION
**Pull request by LJ171**

### CODE REVIEW

File: kge/model/transr.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        out = torch.matmul(projection_matrix, ent_emb)
+
+        return out.reshape(-1, d)
+
+    def score_emb(self, s_emb, p_emb, o_emb, combine: str):
+        test = torch.chunk(p_emb, 129, dim=1)
+
+        rel_emb = test[0]
+        projection_matrix = torch.cat(test[-(len(test)-1):])
```
Comment (LJ171) [inline kge/model/transr.py]: The way the projection matrix is obtained feels unclean at best, what's your opinion about that? (Also see line 116)

Comment (samboy83) [reply kge/model/transr.py]: I don't think it is a good idea to use specific number 64.5 in the transr_set_relation_embedder_dim method and 129 in torch.chunk(p_emb, 129, dim=1).
What we want is entity embedding (dim D), relation embedding (dim K) and projection matrix (dim K X D).
So, I think simply first change the dim of relation from K to K + K * D (= K * (1 + D)) in the transr_set_relation_embedder_dim method
and then divide p_emb into rel_emb and projection_matrix by using slicing like
```python
rel_emb = p_emb[:,K:]
projection_matrix = p_emb[:,:K]
```

What do you think about this?

Comment (LJ171) [reply kge/model/transr.py]: Yeah the hard-coded 64.5 & 129 is definitely not the way to go. I already changed it up a bit, see https://github.com/LJ171/kge/pull/2/commits/6688e828f3816000b5a4e4eb944a5ffda33a1420. It didn't occurred to me before that I can use 'config.get_default(rel_emb_conf_key + ".dim")' to retrieve the original k value and then apply the calculation as proposed by you. Thanks for the help 👍 implemented with https://github.com/LJ171/kge/pull/2/commits/500c0f7f01cba47a33f57a7ebb65ba08a1c6e792.

File: kge/model/transr.py [hunk 2]
```diff
@@ -98,24 +99,23 @@ def __init__(
...
     ent_emb_conf_key = rel_emb_conf_key.replace(
         "relation_embedder", "entity_embedder"
     )
     if ent_emb_conf_key == rel_emb_conf_key:
         raise ValueError(
             "Cannot determine relation embedding size. "
-            "Please set it manually to 'd * (d/2 + 0.5)'. "
+            "Please set it manually to 'k * (1 + d)'. "
             "For d = entity embedder dimensionality. "
+            "For k = relation embedder dimensionality. "
```
Comment (samboy83) [inline kge/model/transr.py:113]: Is this part necessary?

File: kge/model/transr.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    ent_emb_conf_key = rel_emb_conf_key.replace(
+        "relation_embedder", "entity_embedder"
+    )
+    if ent_emb_conf_key == rel_emb_conf_key:
+        raise ValueError(
+            "Cannot determine relation embedding size. "
+            "Please set it manually to 'k * (1 + d)'. "
+            "For d = entity embedder dimensionality. "
+            "For k = relation embedder dimensionality. "
+        )
```
Comment (samboy83) [inline kge/model/transr.py]: Is this part necessary? 
dim_d is from entity_embedder.dim and dim_k is relation_embedder.dim 
and new_dim_k is calculated by dim_d and dim_k. 
So I don't think that we need to check this.

Comment (LJ171) [reply kge/model/transr.py]: I'm not sure whether it is necessary (I guess not) but since TransH had this check I added it: https://github.com/LJ171/kge/blob/d0634774b1b29afc739454737aa105dec6cb60b2/kge/model/transh.py#L156
Should I remove it?

Comment (samboy83) [reply kge/model/transr.py]: This part is used in RESCAL and TransH, and the reason is their default relation embedding dimenstion is -1 which is to determine relation dimension automatically.
But in TransR, the relation embedding dimenstion can't be determined automatically (so default value is not -1), so I think this part is unnecessary.

Comment (LJ171) [reply kge/model/transr.py]: Ahh, okay make sense, then I will remove it.

Comment (LJ171) [reply kge/model/transr.py]: Addressed with https://github.com/LJ171/kge/pull/2/commits/f3ca3017653f7509a38c643c12e0c24840fe278d.

**Comment by samboy83** [comment]
See my comment https://github.com/LJ171/kge/pull/3#discussion_r628873219
If it changes like below, it seems to work well
```python
if not init_for_load_only:
    transr_set_relation_embedder_dim(
        config, dataset, self.configuration_key + ".relation_embedder"
    )
```

**Comment by LJ171** [comment]
> See my comment [#3 (comment)](https://github.com/LJ171/kge/pull/3#discussion_r628873219)
> If it changes like below, it seems to work well
> 
> ```python
> if not init_for_load_only:
>     transr_set_relation_embedder_dim(
>         config, dataset, self.configuration_key + ".relation_embedder"
>     )
> ```

Nice observation 👍 I changed it and now it works. I tried it with TransH too. Do you got a clue why TransH does not need this even though they also change the dimensions?

**Comment by samboy83** [comment]
> 
> Nice observation 👍 I changed it and now it works. I tried it with TransH too. Do you got a clue why TransH does not need this even though they also change the dimensions?

That's because TransH (and RESCAL, too) changes dimension only when the relation dimension is -1 (```if dim < 0: ```).
After the dimension changes, the relation dimension becomes > 0, so the dimension doesn't change anymore.
By the way, why did you make the transfer function static?

**Comment by LJ171** [comment]
> > Nice observation 👍 I changed it and now it works. I tried it with TransH too. Do you got a clue why TransH does not need this even though they also change the dimensions?
> 
> That's because TransH (and RESCAL, too) changes dimension only when the relation dimension is -1 (`if dim < 0: `).
> After the dimension changes, the relation dimension becomes > 0, so the dimension doesn't change anymore.
> By the way, why did you make the transfer function static?

Ahhh okay.
I made it static mainly because of Clean Code and it saves some memory since python doesn't have to initial the bound-method but this is minor.
